docs/generic-constraints.md

```diff
@@ -0,0 +1,142 @@
+# Generic Constraints
+## Summary
+Introduce syntax to constrain the type of a generic.
+## Motivation
+Luau currently does not provide a way in order to constrain the type of a generic.
+```luau
+local qux = { foo = 10, bar = "string" }
+
+local function getProperty<T>( object: T, key: keyof<T> ): index<T, keyof<T>>
+	return object[key]
+end
+
+local foo = getProperty( qux, "foo" )
+-- foo: number | string
+local bar = getProperty( qux, "bar" )
+-- bar: number | string
+```
+Type interference believes that either value could be a number or string as `keyof<T>` is a union type of either foo or bar.
+```luau
+local function callbackProperty<T>( object: T, key: keyof<T>, callback: (property: index<T, keyof<T>>) -> () )
+....
+```
+It is impossible to tell whether or not `key` is the same variable being used to index `T` in the callback, and thus Luau infers the type of `property` to be `number | string`.
+```luau
+local function callbackProperty<T>( object: T, key: keyof<T>, callback: (property: index<T, K>) -> () )
+....
+```
+This does not work in Luau.
+> Ideally, this behaviour should work out of the box, but, could be solved with updates to the type engine.
+
+## Design
+### Syntax
+There are two new options for syntax that come with this generic constraints.
+1. Treat the generic parameter as a variable--annotate it as declaring a new variable.
```

Comment:
> "Treat the generic parameter as a variable--annotate it as declaring a new variable."

It actually _is_ a variable, but it ranges over types instead of values, so `X: T` is a natural consequence from that.